src/diffusers/models/embeddings/combined.py

```diff
@@ -0,0 +1,173 @@
+
+        self.use_additional_conditions = use_additional_conditions
+        if use_additional_conditions:
+            self.additional_condition_proj = Timesteps(num_channels=256, flip_sin_to_cos=True, downscale_freq_shift=0)
+            self.resolution_embedder = TimestepEmbedding(in_channels=256, time_embed_dim=size_emb_dim)
+            self.aspect_ratio_embedder = TimestepEmbedding(in_channels=256, time_embed_dim=size_emb_dim)
+
+    def forward(self, timestep, resolution, aspect_ratio, batch_size, hidden_dtype):
+        timesteps_proj = self.time_proj(timestep)
+        timesteps_emb = self.timestep_embedder(timesteps_proj.to(dtype=hidden_dtype))  # (N, D)
+
+        if self.use_additional_conditions:
+            resolution_emb = self.additional_condition_proj(resolution.flatten()).to(hidden_dtype)
+            resolution_emb = self.resolution_embedder(resolution_emb).reshape(batch_size, -1)
+            aspect_ratio_emb = self.additional_condition_proj(aspect_ratio.flatten()).to(hidden_dtype)
+            aspect_ratio_emb = self.aspect_ratio_embedder(aspect_ratio_emb).reshape(batch_size, -1)
+            conditioning = timesteps_emb + torch.cat([resolution_emb, aspect_ratio_emb], dim=1)
+        else:
+            conditioning = timesteps_emb
+
+        return conditioning
+
+
+class HunyuanCombinedTimestepTextSizeStyleEmbedding(nn.Module):
+    def __init__(self, embedding_dim, pooled_projection_dim=1024, seq_len=256, cross_attention_dim=2048):
+        super().__init__()
+        from .image_text import PixArtAlphaTextProjection
+        from .others import HunyuanDiTAttentionPool
+        from .timestep import TimestepEmbedding, Timesteps
+
+        self.time_proj = Timesteps(num_channels=256, flip_sin_to_cos=True, downscale_freq_shift=0)
+        self.timestep_embedder = TimestepEmbedding(in_channels=256, time_embed_dim=embedding_dim)
+
+        self.pooler = HunyuanDiTAttentionPool(
+            seq_len, cross_attention_dim, num_heads=8, output_dim=pooled_projection_dim
+        )
+        # Here we use a default learned embedder layer for future extension.
+        self.style_embedder = nn.Embedding(1, embedding_dim)
+        extra_in_dim = 256 * 6 + embedding_dim + pooled_projection_dim
+        self.extra_embedder = PixArtAlphaTextProjection(
+            in_features=extra_in_dim,
+            hidden_size=embedding_dim * 4,
+            out_features=embedding_dim,
+            act_fn="silu_fp32",
+        )
+
+    def forward(self, timestep, encoder_hidden_states, image_meta_size, style, hidden_dtype=None):
+        timesteps_proj = self.time_proj(timestep)
+        timesteps_emb = self.timestep_embedder(timesteps_proj.to(dtype=hidden_dtype))  # (N, 256)
+
+        # extra condition1: text
+        pooled_projections = self.pooler(encoder_hidden_states)  # (N, 1024)
+
+        # extra condition2: image meta size embdding
+        image_meta_size = get_timestep_embedding(image_meta_size.view(-1), 256, True, 0)
```

Comment:
Does this imply that the embedding size can change based on the image size? Am I understanding that correctly?

Reply:
I didn't introduce anything new. It's from:
https://github.com/huggingface/diffusers/blob/49979753e14e7fed61bebd5700d4dbd1b14a6ffc/src/diffusers/models/embeddings.py#L760

Reply:
this is the same idea as in SDXL actually
maybe we can refactor with https://github.com/huggingface/diffusers/blob/49979753e14e7fed61bebd5700d4dbd1b14a6ffc/src/diffusers/models/embeddings.py#L388

Reply:
Sure. Let's maybe keep the embedding-specific refactorings for a future PR and use this PR to tear apart `embeddings.py` into type-specific refactorings only? Easier for reviewers IMO.

Reply:
it is related to this though https://github.com/huggingface/diffusers/pull/8722#discussion_r1661307759

Reply:
ok to do it in a different PRs - but I think we can do this first (happy to take this one too since it's related to hunyuan) https://github.com/huggingface/diffusers/pull/8722#discussion_r1661289218?

Reply:
Could you elaborate? I don’t understand how it’s related.

Reply:
Sure. Let’s have it first then